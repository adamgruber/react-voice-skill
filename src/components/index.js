/* eslint import/prefer-default-export: 0 */
import VoiceSkill from './VoiceSkill';

export { VoiceSkill };
