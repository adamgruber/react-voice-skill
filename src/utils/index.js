import Alexa from 'alexa-sdk';
import { EventEmitter } from 'events';
import VoiceSkill from '../components/VoiceSkill';
import IntentHandler from '../components/IntentHandler';
import Response from '../components/Response';
import Speak from '../components/Speak';
import Card from '../components/Card';

let ROOT_NODE = null;
let RESPONSE_ROOT_NODE = null;

/**
 * Updates the ref to ROOT_NODE / RESPONSE_ROOT_NODE
 *
 * @param {*} rootNode root instance
 */
function getHostContext(rootNode) {
  console.log('getHostContext');
  if (rootNode instanceof EventEmitter) {
    return ROOT_NODE = rootNode;
  }

  return RESPONSE_ROOT_NODE = rootNode;
}

/**
 * Creates an element for a skill
 *
 * @param {string} type Element type
 * @param {Object} props Component props
 * @returns {Object}
 */
function createElement(type, props) {
  console.log(`Create Element: ${type}\n`);
  // console.log(props);
  // console.log(ROOT_NODE);

  const components = {
    ROOT: () => {
      const alexa = Alexa.handler(props.event, props.context);
      alexa.appId = props.appId;
      return alexa;
    },
    RESPONSE_ROOT: () => props.handlerContext,

    VOICESKILL: () => new VoiceSkill(null, { ...props, $$type: 'VoiceSkill' }),
    INTENTHANDLER: () => new IntentHandler(ROOT_NODE, { ...props, $$type: 'IntentHandler' }),

    RESPONSE: () => new Response(RESPONSE_ROOT_NODE, { ...props, $$type: 'Response' }),
    SPEAK: () => new Speak(RESPONSE_ROOT_NODE, { ...props, $$type: 'Speak' }),
    CARD: () => new Card(RESPONSE_ROOT_NODE, { ...props, $$type: 'Card' }),
  };

  const elementCreator = components[type];

  if (elementCreator) {
    return elementCreator();
  } else {
    throw new Error(`Unknown element: ${type}`);
  }
}

export {
  createElement,
  getHostContext,
};
