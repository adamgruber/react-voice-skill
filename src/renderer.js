/* eslint no-param-reassign: 0 */
import ReactFiberReconciler from 'react-reconciler';
import { createElement, getHostContext } from './utils';

const noop = () => {};

const AlexaRenderer = ReactFiberReconciler({
  appendInitialChild(parentInstance, child) {
    console.log('appendInitialChild');
    if (parentInstance.appendChild) {
      parentInstance.appendChild(child);
    } else {
      parentInstance.skill = child;
    }
  },

  createInstance(type, props, internalInstanceHandle) {
    return createElement(type, props, internalInstanceHandle);
  },

  createTextInstance(text) {
    return text;
  },

  finalizeInitialChildren() {
    return false;
  },

  getPublicInstance(instance) {
    return instance;
  },

  prepareForCommit: noop,

  prepareUpdate() {
    return true;
  },

  resetAfterCommit: noop,
  resetTextContent: noop,

  getRootHostContext(rootInstance) {
    return getHostContext(rootInstance);
  },

  getChildHostContext() {
    return {};
  },

  shouldSetTextContent() {
    return false;
  },

  now: noop,

  useSyncScheduling: true,

  mutation: {
    appendChild(parentInstance, child) {
      console.log('mutation appendInitialChild');
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        parentInstance.skill = child;
      }
    },

    appendChildToContainer(parentInstance, child) {
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        switch (child.type) {
          case 'VoiceSkill':
            parentInstance.skill = child;
            break;

          case 'Response':
            parentInstance.res = child;
            break;

          default:
            // ignore
        }
      }
    },

    removeChild: noop,
    removeChildFromContainer: noop,
    insertBefore: noop,
    commitUpdate: noop,
    commitMount: noop,
    commitTextUpdate: noop,
  },
});

/**
 * @param  {ReactElement} element
 * @param {object} options
 * @param {string} options.appId Alexa App ID
 * @param {object} options.event Lambda request event
 * @param {object} options.context Lambda request context
 * @return {object}
 */
const renderToAlexa = (element, { appId, event, context }) => {
  const container = createElement('ROOT', { appId, event, context });
  const node = AlexaRenderer.createContainer(container);
  AlexaRenderer.updateContainer(element, node, null);
  const handlers = container.skill.render();
  console.log(handlers);
  container.registerHandlers(handlers);
  return container.execute();
};

/**
 * @param  {ReactElement} element
 * @param {object} handlerContext ASK SDK Handler Context
 * @return {object}
 */
const renderAlexaResponse = (element, handlerContext) => {
  const container = createElement('RESPONSE_ROOT', { handlerContext });
  const node = AlexaRenderer.createContainer(container);
  AlexaRenderer.updateContainer(element, node, null);

  // Get our response object from rendering components
  const response = container.res.render();
  console.log(response);

  // Parse the response object to build our Alexa response
  Object.keys(response).forEach((type) => {
    const res = response[type];
    switch (type) {
      case 'Card':
        container.response.cardRenderer(res.title, res.content, res.imgSrc);
        break;

      case 'Speak':
        container.response.speak(res.content);
        if (res.listen) {
          container.response.listen(res.listen);
        }
        break;

      default:
        // ignore
    }
  });
  return container.emit(':responseReady');
};

export {
  AlexaRenderer,
  renderToAlexa,
  renderAlexaResponse,
};
