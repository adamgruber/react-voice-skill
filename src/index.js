import { renderToVoiceSkill, renderToAlexa } from './renderer';

const Card = 'CARD';
const IntentHandler = 'INTENTHANDLER';
const Response = 'RESPONSE';
const Speak = 'SPEAK';
const VoiceSkill = 'VOICESKILL';

export {
    Card,
    IntentHandler,
    renderToAlexa,
    renderToVoiceSkill,
    Response,
    Speak,
    VoiceSkill,
};
