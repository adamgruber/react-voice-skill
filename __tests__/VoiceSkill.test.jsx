import React from 'react';
import { VoiceSkill, IntentHandler, Response } from '../src';
import VoiceSkillComp from '../src/components/VoiceSkill';
import render from './testRenderer';
import { IntentRequest, APP_ID } from '../examples/fixtures';
import context from '../examples/sample_context';

describe('VoiceSkill', () => {
  let App;
  let event;

  const renderApp = () => render(<App />, { appId: APP_ID, event, context });

  beforeEach(() => {
    event = IntentRequest({
      intent: {
        name: 'LaunchRequest',
      },
    });
  });

  it('should render element', () => {
    App = () => <VoiceSkill />;
    expect(renderApp()).toMatchSnapshot();
  });

  it('should render element with children', () => {
    App = () => (
      <VoiceSkill>
        <IntentHandler intent="LaunchRequest" />
      </VoiceSkill>
    );
    expect(renderApp()).toMatchSnapshot();
  });

  it('should throw with invalid children', () => {
    App = () => (
      <VoiceSkill>
        <Response />
      </VoiceSkill>
    );
    expect(() => renderApp()).toThrow('Response is not a valid child of VoiceSkill');
  });

  describe('getValidChildren', () => {
    let comp;

    beforeEach(() => {
      comp = new VoiceSkillComp();
    });

    it('should return array of valid child types', () => {
      expect(comp.getValidChildren()).toEqual(['IntentHandler']);
    });
  });
});
