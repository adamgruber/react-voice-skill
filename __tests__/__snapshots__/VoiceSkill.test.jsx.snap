// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoiceSkill should render element 1`] = `
Object {
  "component": VoiceSkill {
    "children": Array [],
    "props": Object {
      "$$type": "VoiceSkill",
    },
    "root": null,
    "type": "VoiceSkill",
  },
  "rendered": Object {},
}
`;

exports[`VoiceSkill should render element with children 1`] = `
Object {
  "component": VoiceSkill {
    "children": Array [
      IntentHandler {
        "children": Array [],
        "props": Object {
          "$$type": "IntentHandler",
          "intent": "LaunchRequest",
        },
        "root": AlexaRequestEmitter {
          "_events": Object {
            ":ask": [Function],
            ":askWithCard": [Function],
            ":askWithLinkAccountCard": [Function],
            ":askWithPermissionCard": [Function],
            ":confirmIntent": [Function],
            ":confirmIntentWithCard": [Function],
            ":confirmSlot": [Function],
            ":confirmSlotWithCard": [Function],
            ":delegate": [Function],
            ":elicitSlot": [Function],
            ":elicitSlotWithCard": [Function],
            ":responseReady": [Function],
            ":saveState": [Function],
            ":saveStateError": [Function],
            ":tell": [Function],
            ":tellWithCard": [Function],
            ":tellWithLinkAccountCard": [Function],
            ":tellWithPermissionCard": [Function],
            "AlexaSkillEvent.SkillAccountLinked": [Function],
            "AlexaSkillEvent.SkillDisabled": [Function],
            "AlexaSkillEvent.SkillEnabled": [Function],
            "AlexaSkillEvent.SkillPermissionAccepted": [Function],
            "AlexaSkillEvent.SkillPermissionChanged": [Function],
          },
          "_eventsCount": 23,
          "_maxListeners": Infinity,
          "domain": null,
          "skill": [Circular],
        },
        "type": "IntentHandler",
      },
    ],
    "props": Object {
      "$$type": "VoiceSkill",
      "children": <INTENTHANDLER
        intent="LaunchRequest"
      />,
    },
    "root": null,
    "type": "VoiceSkill",
  },
  "rendered": Object {
    "LaunchRequest": [Function],
  },
}
`;
